### DISCUSSION TITLE
Using torchmetrics metrics with Chemprop for imbalanced classification

### DISCUSSION THREAD

**Discussion by MarcS18** [chemprop/chemprop#1344] [Show and tell]
I thought I would note down here how to use metrics from torchmetrics during Chemprop training workflows as I've come across this a couple of times now and always forget how to do it. When performing classification tasks on imbalanced data the base evaluation metrics can be insufficient if the classes are highly imbalanced. In these cases metrics such as balanced accuracy, macro-F1 score/Recall/Precision are more desirable. Below is an example of how to wrap and use metrics from torchmetrics during Chemprop training and evaluation.

```python
import torchmetrics
from chemprop.nn.metrics import MetricRegistry, ClassificationMixin

@MetricRegistry.register("bal-acc")
class BalancedAccuracy(ClassificationMixin, torchmetrics.classification.MulticlassAccuracy):
    def __init__(self, *args, **kwargs):
        super(torchmetrics.classification.MulticlassAccuracy, self).__init__(*args, **kwargs)

@MetricRegistry.register("macro-f1")
class MacroF1Score(ClassificationMixin, torchmetrics.classification.MulticlassF1Score):
    def __init__(self, *args, **kwargs):
        super(torchmetrics.classification.MulticlassF1Score, self).__init__(*args, **kwargs)
```

The process is the same for similar metrics. This new class can then be called the same as any other Chemprop metric allowing all the base metrics arguments:

```python
# For binary classification
# We must pass a value for beta to the F1Score, for base F1Score beta should equal 1.0.
metrics = [BalancedAccuracy(num_classes=2, average="macro"), MacroF1Score(1.0, num_classes=2, average="macro")] 
```

We can then create the MPNN as normal:

```python
mpnn = MPNN(mp, agg, ffn, metrics=metrics)
```

Hopefully this is helpful for others!
